# 🚗 Car Price Prediction – Regression Model

This project focuses on predicting the price of a car based on its features and specifications using machine learning regression techniques.

The dataset contains information about different cars, such as brand, model, year, mileage, fuel type, transmission, engine characteristics, and other relevant features. 

The goal is to analyze these features and build a machine learning model that can estimate the expected price of a car.

#  Data Exploration

* 📊 Number of rows and columns
* 🔤 Data types
* ❌ Missing values
* 🎯 Target variable: priceUSD
* 🔢 Numerical columns
* ⚠️ Extreme values and outliers
* 📈 Data distribution and potential anomalies

# 📋 Step 1: 
* Import pandas library, Path
* Read and load the dataset
* Info about columns

In [67]:
from pathlib import Path
import pandas as pd

DATA_PATH = Path("../data/cars.csv")

try:
    df = pd.read_csv(DATA_PATH)
    print("---- ✅ Dataset loaded successfully.")
except FileNotFoundError:
    print("Error: The dataset file was not found.")
except pd.errors.EmptyDataError:
    print("Error: The dataset is empty.")
except pd.errors.ParserError:
    print("Error: Failed to parse the dataset.")
except Exception as e:
    print(f"An unexpected error occurred: {e}")

df.columns


---- ✅ Dataset loaded successfully.


Index(['make', 'model', 'priceUSD', 'year', 'condition', 'mileage(kilometers)',
       'fuel_type', 'volume(cm3)', 'color', 'transmission', 'drive_unit',
       'segment'],
      dtype='str')

The columns need to be standardize in data_cleaning.

# 📋 Step 2: 
* 📊 Number of rows and columns
* First look at the data

In [68]:
print("Number of rows and columns:",df.shape)
df.sample(5,random_state=42)

Number of rows and columns: (56244, 12)


,make,model,priceUSD,year,condition,mileage(kilometers),fuel_type,volume(cm3),color,transmission,drive_unit,segment
19270,mitsubishi,carisma,2050,1999,with mileage,250000.0,petrol,1800.0,blue,mechanics,front-wheel drive,M
25927,ford,fusion,4500,2006,with mileage,160000.0,petrol,1400.0,burgundy,mechanics,front-wheel drive,M
23388,mercedes-benz,e-klass,3500,1999,with mileage,485000.0,diesel,2900.0,blue,mechanics,rear drive,E
53189,bmw,x3,21000,2013,with mileage,159000.0,diesel,2000.0,black,auto,NaN,J
34058,renault,megane,3990,2002,with mileage,331700.0,diesel,1900.0,other,mechanics,front-wheel drive,C


# 📋 Step 3:
 
* 🔤 Data types
* ❌ Missing values

In [69]:
print("🔍Main Info.")
print("-"*40)
print(df.info())
print("-"*40)
print("----Count Missing Values in dataset:")
print(df.isna().sum())

missing_like_values = ["NaN", "nan", "NULL", "null", "None", "none", "", " "]
 
for column in df.columns:
    if df[column].dtype == "object":
        count = df[column].astype(str).str.strip().isin(missing_like_values).sum()
         
        if count > 0:
            print(f"{column}:{count} missing values.")
    else:  
            print(f"🔍🧩✅  {column}: Doesn't contain any values from missing_values.")



🔍Main Info.
----------------------------------------
<class 'pandas.DataFrame'>
RangeIndex: 56244 entries, 0 to 56243
Data columns (total 12 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   make                 56244 non-null  str    
 1   model                56244 non-null  str    
 2   priceUSD             56244 non-null  int64  
 3   year                 56244 non-null  int64  
 4   condition            56244 non-null  str    
 5   mileage(kilometers)  56244 non-null  float64
 6   fuel_type            56244 non-null  str    
 7   volume(cm3)          56197 non-null  float64
 8   color                56244 non-null  str    
 9   transmission         56244 non-null  str    
 10  drive_unit           54339 non-null  str    
 11  segment              50953 non-null  str    
dtypes: float64(2), int64(2), str(8)
memory usage: 5.1 MB
None
----------------------------------------
----Count Missing Values in dataset:
make    

### 🔍 Main Info

* 📊 The dataset contains **56,244 rows and 12 columns**.
* 🔢 There are **4 numerical columns** and **8 categorical columns**.
* 🎯 The target variable is **price_usd**.
* ❌ **volume_cm3** contains **47 missing values**.
* ❌ **drive_unit** contains **1,905 missing values**.
* ❌ **segment** contains **5,291 missing values**.
* ✅ The remaining columns have **no missing values**.
* 💾 The dataset uses approximately **5.1 MB of memory**.





# 📋 Step 4: 
* 📊  Numerical columns
* 📊  Categorical columns

In [70]:
numeric_columns = [
    "priceUSD",
    "year",
    "mileage(kilometers)",
    "volume(cm3)"
]

categorical_columns = [
    "make",
    "model",
    "condition",
    "fuel_type",
    "color",
    "transmission",
    "drive_unit",
    "segment"
]
print(df[numeric_columns].dtypes)
print(df[numeric_columns].describe())

for column in categorical_columns:
    print(column)
    print(df[column].nunique())
    print("-" * 40)
    
    
for column in categorical_columns:
    print(column)
    print(df[column].value_counts(dropna=False))
    print("-" * 40)

priceUSD                 int64
year                     int64
mileage(kilometers)    float64
volume(cm3)            float64
dtype: object
            priceUSD          year  mileage(kilometers)   volume(cm3)
count   56244.000000  56244.000000         5.624400e+04  56197.000000
mean     7415.456440   2003.454840         2.443956e+05   2104.860615
std      8316.959261      8.144247         3.210307e+05    959.201633
min        48.000000   1910.000000         0.000000e+00    500.000000
25%      2350.000000   1998.000000         1.370000e+05   1600.000000
50%      5350.000000   2004.000000         2.285000e+05   1996.000000
75%      9807.500000   2010.000000         3.100000e+05   2300.000000
max    235235.000000   2019.000000         9.999999e+06  20000.000000
make
96
----------------------------------------
model
1034
----------------------------------------
condition
3
----------------------------------------
fuel_type
3
----------------------------------------
color
13
----------------

### Conclusion – Categorical Features 🧩


The categorical features contain different numbers of unique values. make has 96 unique values, while model has 1,034, indicating high cardinality. 
The other features have fewer categories: condition has 3, fuel_type 3, color 13, transmission 2, drive_unit 4, and segment 8 unique values, drive_unit and segment also contain missing values.
Current data types are appropriate for the dataset.
Further inspection in columns: priceUSD, mileage(kilometers), volume(cm), model...

# 📋 Step 5:
* 🎯 Target variable: priceUSD -- check values

In [71]:
print(df['priceUSD'].describe())

count     56244.000000
mean       7415.456440
std        8316.959261
min          48.000000
25%        2350.000000
50%        5350.000000
75%        9807.500000
max      235235.000000
Name: priceUSD, dtype: float64


⚠️ The maximum price of $235,235 is exceptionaly compared the median price of $5,350. 

This value should be investigated further to determine whether it represents a genuine high-priced vehicle or a potential outlier/data anomaly.

# 📋 Step 6: Info about priceUSD
* 📈 Data distribution and potential anomalies
* ⚠️ Extreme values and outliers


In [72]:
print("🔍🚗Car with the Highest Price.")
print(df.loc[df['priceUSD'].idxmax()])
print("-"*40)
print("🔍🚗Car with the lowest Price.")
print(df.loc[df['priceUSD'].idxmin()])

🔍🚗Car with the Highest Price.
make                           bentley
model                   continental-gt
priceUSD                        235235
year                              2018
condition                 with mileage
mileage(kilometers)            29000.0
fuel_type                       petrol
volume(cm3)                     6000.0
color                            white
transmission                      auto
drive_unit             all-wheel drive
segment                            NaN
Name: 20787, dtype: object
----------------------------------------
🔍🚗Car with the lowest Price.
make                           zaz
model                          968
priceUSD                        48
year                          1910
condition              with damage
mileage(kilometers)        25000.0
fuel_type                   petrol
volume(cm3)                 1100.0
color                        green
transmission             mechanics
drive_unit              rear drive
segment             

### 🔎 Maximum Price Investigation

The highest price in the dataset is **$235,235**. After investigating the corresponding record, I found that it belongs to a **2018 Bentley Continental GT** with **29,000 km** mileage and a **6000 cm³** petrol engine.

The value appears to be plausible for this type of luxury vehicle, so it should not be automatically treated as an incorrect value. However, it represents an extreme value compared to most cars in the dataset and may influence the regression model.

### 🔎 Minimum Price Investigation

The minimum price is $48 for a damaged ZAZ 968. The year 1910 appears to be anomalous, but since the correct year cannot be confirmed with certainty, the original value will be retained for now.

# 📋 Step 7: Info about mileage(kilometers)
* 📈 Data distribution and potential anomalies
* ⚠️ Extreme values and outliers


In [73]:
print("🔍🚗Car with the Highest Mileage(kilometers).")
print(df.loc[df['mileage(kilometers)'].idxmax()])
print("-"*40)
print("🔍🚗Car with the lowest Mileage(kilometers).")
print(df.loc[df['mileage(kilometers)'].idxmin()])

🔍🚗Car with the Highest Mileage(kilometers).
make                                audi
model                                 80
priceUSD                             500
year                                1986
condition                   with mileage
mileage(kilometers)            9999999.0
fuel_type                         petrol
volume(cm3)                       1800.0
color                             silver
transmission                   mechanics
drive_unit             front-wheel drive
segment                                D
Name: 1394, dtype: object
----------------------------------------
🔍🚗Car with the lowest Mileage(kilometers).
make                               mazda
model                                  3
priceUSD                            5500
year                                2007
condition                   with mileage
mileage(kilometers)                  0.0
fuel_type                         petrol
volume(cm3)                       1600.0
color                     

### 🔍🚗 Mileage Analysis

* 🚨 **Highest mileage:** 9,999,999 km – Audi 80 (1986). This value appears to be an extreme data anomaly and should be investigated further.
* 🔍 **Lowest mileage:** 0 km – Mazda 3 (2007). This value may be valid, but it could also represent missing or unknown mileage and should be investigated.


🔍🚗 Further Investigation -------

* Count total anomaly values.
* Display anomaly rows.

In [74]:
print("Total number of rows with mileage in (9999999.0).")
print((df['mileage(kilometers)'] == 9999999.0).sum())
print("Total number of rows with mileage in (0.0)")
print((df['mileage(kilometers)'] == 0).sum())

Total number of rows with mileage in (9999999.0).
22
Total number of rows with mileage in (0.0)
190


In [75]:
print(df[df['mileage(kilometers)'] == 9999999.0])
print("-"*40)
print(df[df['mileage(kilometers)'] == 0])

                make     model  priceUSD  year     condition  \
1394            audi        80       500  1986  with mileage   
1854            audi       100       900  1989  with mileage   
5783         peugeot       806      4100  2000  with mileage   
6114        lada-vaz      2104       600  1990  with mileage   
6465        lada-vaz      2108       295  1990  with mileage   
6684             raf      2203      1000  1991  with mileage   
10032            bmw  5-seriya      2000  1985  with mileage   
10316            bmw  7-seriya      4650  1998  with mileage   
11699           audi        a4      3000  1996  with mileage   
16705           fiat     brava      1100  1997  with mileage   
21525          honda      cr-v      7300  2005  with mileage   
23161  mercedes-benz   e-klass      1850  1989  with mileage   
30631           opel    kadett       250  1985     for parts   
31277        renault    laguna      1400  1995  with mileage   
32444         nissan    laurel       550

🔍🚗 Further Investigation – Extreme Mileage

After investigating all 22 vehicles with a mileage of 9,999,999 km, the value appears across different brands, models, years, prices, and vehicle conditions.
Since such a mileage value is unrealistic and occurs repeatedly across unrelated vehicles, it is most likely a placeholder or data entry anomaly rather than the actual mileage.
Therefore, these values should be treated as invalid mileage values during data preprocessing.

🔍🚗 Further Investigation – Zero Mileage

There are 190 vehicles with a mileage of 0 km. These values occur across different brands, models, years, and vehicle conditions.
Since zero mileage is unusual for used vehicles, these records require further investigation. However, there is not enough evidence to conclude that all of them are incorrect, so the values will be retained for now and considered during data preprocessing.

In [76]:
df[df['mileage(kilometers)'] == 0]['condition'].value_counts()

condition
with mileage    174
for parts        14
with damage       2
Name: count, dtype: int64

🔍🚗 Zero Mileage Conclusion

Among the 190 vehicles with 0 km mileage:

* 🚗 174 have the condition with mileage
* 🔧 14 have the condition for parts
* ⚠️ 2 have the condition with damage

The 174 vehicles listed as with mileage with 0 km are particularly unusual and may indicate missing or incorrectly recorded mileage. However, without reliable evidence of the actual mileage, these values will not be changed at this stage.

# 📋 Step 8:
* 🔍🚗 Volume(cm) analysis

In [77]:
df[df["volume(cm3)"] >= 10000][
    ["make", "model", "volume(cm3)","priceUSD"]
].sort_values("volume(cm3)", ascending=False)

,make,model,volume(cm3),priceUSD
81,mazda,3,20000.0,5300
354,mazda,6,20000.0,7500
1537,audi,80,20000.0,2900
1768,audi,100,20000.0,2150
4573,peugeot,407,20000.0,4600
...,...,...,...,...
3714,mazda,323,13000.0,1500
6338,lada-vaz,2106,13000.0,1236
6286,lada-vaz,2106,13000.0,530
6065,lada-vaz,2101,12000.0,214


Conclusion

The volume(cm3) column contains several unrealistic values. Values between 10,000 and 20,000 cm³ are clearly suspicious for the car models in this dataset. Therefore, these values should be treated as invalid and replaced with NaN before further data preprocessing.

# 📋 Step 9:
* 🚗 Model Analysis

In [78]:

print(df['model'].value_counts())
print("-"*40)
print("The number of model unique categories:",df['model'].nunique())

model
passat      2086
5-seriya    1476
a6          1276
golf        1070
astra       1013
            ... 
xb             1
xc40           1
xjs            1
xt5            1
z3             1
Name: count, Length: 1034, dtype: int64
----------------------------------------
The number of model unique categories: 1034


### 🚗 Model  Analysis Conclusion

The `model` column contains **1,034 unique categories**. Some models occur frequently, while others appear only once or a few times in the dataset.

This high cardinality should be considered during preprocessing, as applying One-Hot Encoding directly would create a large number of additional features.
